Index geopoint case properties in case search index

esoergel opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
No user-facing behavior changes yet

## Technical Summary
https://dimagi-dev.atlassian.net/browse/USH-1753

Based on the testing described in this draft PR: https://github.com/dimagi/commcare-hq/pull/31266
When a project enables the (new) feature flag `CASE_SEARCH_SMART_TYPES`, any case properties set to type `gps` in the data dictionary will be sent to ES with an additional `geopoint_value` field, allowing them to be used for `geo_distance` queries in the future.

I'll separately add a management command to allow us to manually reindex a domain, which I imagine we'll want to do eventually in support of this change.

See the Migrations section below for rollout.

## Feature Flag
This introduces a new feature flag:
> Intelligently index specific case properties using the data dictionary

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
I tested this out on staging with a normal deploy and had no issues.  I then tested out the change and confirmed that new data got the field indexed as expected, and was queryable.  ES metrics in datadog didn't show anything notable.

### Automated test coverage

Included

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Migrations

This does change the case search mapping via django migration.  It should be able to go live during a normal deploy without disrupting things.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Add rudimentary geopoint indexing

esoergel committed:
Use data dictionary to identify GPS properties

TODO cache!

esoergel committed:
Cache data dictionary lookup

esoergel committed:
Move doctests to test file

esoergel committed:
Move the parsing method to GeoPoint

This doesn't really need to have anything to do with JsonObject for most
of its usages

esoergel committed:
Allow two-element geopoints

esoergel committed:
Use flexible geopoints for case search properties

esoergel committed:
Make geo_distance query function

snopoke commented:
> Can this instead be applied during a normal deploy?
How will this rollout to third parties?
Should I make a django migration that calls update_es_mapping?

There is precedent for updating ES mapping in a deploy: 
https://github.com/dimagi/commcare-hq/blob/a534cdf74633b73d16cc5f54a027e6378571edb5/corehq/apps/app_manager/migrations/0007_add_linked_app_fields_to_es.py

I would feel more comfortable making that migration if we had a test to verify its safety. I think that could be done but would require creating a copy of the previous mapping file for use in the test.

esoergel committed:
This is a USH flag

proteusvacuum commented on corehq/toggles/__init__.py:
```diff
@@ -932,6 +932,13 @@ def _enable_search_index(domain, enabled):
 
+CASE_SEARCH_SMART_TYPES = StaticToggle(
+    'case_search_smart_types',
+    'USH: Intelligently index specific case properties using the data dictionary',
```
Yay to the DD!

proteusvacuum commented on corehq/apps/data_dictionary/util.py:
```diff
@@ -229,3 +229,13 @@ def fields_to_validate(domain, case_type_name):
+
+
+@quickcache(['domain', 'case_type'], timeout=24 * 60 * 60)
+def get_smart_types_by_prop(domain, case_type):
```
Here you are only filtering by `gps`, maybe rename this function to match what it is returning?

esoergel replied:
I'm anticipating this being used for a couple other types in the medium-term future, but I guess it's not hard to change and there's always the likelihood that YAGNI.

proteusvacuum commented on corehq/apps/data_dictionary/util.py:
```diff
@@ -229,3 +229,13 @@ def fields_to_validate(domain, case_type_name):
+    return dict(CaseProperty.objects.filter(
+        case_type__domain=domain,
+        case_type__name=case_type,
+        data_type='gps',
```
nit: might be worth making this a constant, since you reference it in many places.

snopoke replied:
I think this would still be good to do.

esoergel replied:
:+1:

snopoke reviewed (commented):
Generally looks good

esoergel commented:
> There is precedent for updating ES mapping in a deploy:
https://github.com/dimagi/commcare-hq/blob/a534cdf74633b73d16cc5f54a027e6378571edb5/corehq/apps/app_manager/migrations/0007_add_linked_app_fields_to_es.py

Superb, thanks!

> I would feel more comfortable making that migration if we had a test to verify its safety. I think that could be done but would require creating a copy of the previous mapping file for use in the test.

@snopoke that sounds great to me, what did you have in mind?  Something like:
* initialize the index, add some cases
* run the update management command, add more cases
* verify that the mapping change was applied (that it's queryable, I guess)

It'd be nice to also confirm that our deploy process and pillowtop will both handle the change smoothly, but I'm not sure how to do that in a test.

snopoke commented on corehq/pillows/case_search.py:
```diff
@@ -100,11 +104,24 @@ def _get_case_properties(doc_dict):
+            try:
+                prop['geopoint_value'] = GeoPoint.from_string(prop['value'], flexible=True).lat_lon
+            except BadValueError:
+                prop['geopoint_value'] = None
```
I see you commented on that already, fine to leave as is.

esoergel commented on corehq/ex-submodules/couchforms/geopoint.py:
```diff
@@ -0,0 +1,83 @@
+from jsonobject.exceptions import BadValueError
+
+
+class GeoPoint(namedtuple('GeoPoint', 'latitude longitude altitude accuracy')):
```
Simon commented the same.  I suppose this is as good a time as any to update this class and the utility functions.  I'll do a clean-up pass.

esoergel commented on corehq/apps/es/case_search.py:
```diff
@@ -272,6 +272,26 @@ def case_property_missing(case_property_name):
+def case_property_geo_distance(geopoint_property_name, lat, lon, distance):
+    """Filters cases to those within a certain distance of the provided lat/lon
+
+    Distance should be a string like "10km" or "3mi". For options, see:
```
This'd cause an ES error like
```
corehq.apps.es.exceptions.ESError: TransportError(400, 'search_phase_execution_exception', 'For input string: "1000kl"')
```
If it's okay with you I'd like to include validation in the PR that makes use of this filter.  Since the user will be providing the distance in some format, that PR will need user-facing validation, and I thought it'd be easier to do both together.

esoergel commented on corehq/pillows/mappings/case_search_mapping.json:
```diff
@@ -47,6 +47,9 @@
                     }
+                },
+                "geopoint_value": {
+                    "type": "geo_point"
```
This is a JSON file, so I can't include comments, but I could put a note in the pillow code.

esoergel committed:
Use a dataclass

esoergel committed:
Improve exception handling, add more tests

esoergel committed:
Simplify Decimal coersion

The old code did a "hack" where it checked for a specific scientific
notation format and rounded to 0 anything with a negative exponent
smaller larger than 3.  It would fail on any other input provided in
scientific notation with a negative exponent - I'm not sure why.
`Decimal()` accepts scientific notation itself, so this change
simplifies that processing considerably, and introduces support for
numbers between 0.001 and -0.001 if provided in scientific notation.

esoergel committed:
Further clean up parsing code

esoergel committed:
Add comment explaining why we don't duck type it

esoergel commented:
Thanks for the feedback, @snopoke @proteusvacuum .  I cleaned up the `GeoPoint` code at your suggestion.

esoergel committed:
Preserve precision - only round small sci notation

joxl commented:
What's the recommended review process for this PR? Unified or commit-by-commit?

esoergel commented:
@joxl I intended for it to be commit-by-commit, though I think unified should be reasonable too.  If you want to see the changes to the string-to-geopoint parsing in detail, per-commit would be best.

Glad to see you reading this - I was planning on reaching out to you to discuss rollout once the approach here was approved.

snopoke commented on corehq/ex-submodules/couchforms/geopoint.py:
```diff
@@ -52,13 +52,14 @@ def _extract_elements(input_string, flexible):
 
-    if not ret.is_nan() and abs(ret) < Decimal("0.001"):
+    uses_sci_notation = 'e' in n.lower()
+    if uses_sci_notation and abs(ret) < Decimal("0.001"):
```
Why only coerce scientific notation inputs? Surely a normal input below 0.001 is also meaningless?

esoergel replied:
That's what I did at first, but it loses precision.  Some of the existing tests use values of `0.0`, which would be turned into `0`.  It doesn't matter for elasticsearch, where the raw value is also stored, but I'm hesitant to change that behavior for jsonobject.  That is, if you put in `0.0`, that's what you should get out.

snopoke commented on corehq/pillows/case_search.py:
```diff
@@ -114,6 +114,9 @@ def _get_case_properties(doc_dict):
 def _add_smart_types(dynamic_properties, domain, case_type):
+    # Properties are stored in a dict like {"key": "dob", "value": "1900-01-01"}
+    # `value` is a multi-field property that duck types numeric and date values
+    # We can't do that for geo_points in ES v2, as `ignore_malformed` is broken
```
:+1:

snopoke reviewed (approved):
Thanks for the changes and new tests!

Nothing blocking from me.

snopoke commented on corehq/ex-submodules/couchforms/geopoint.py:
```diff
@@ -0,0 +1,83 @@
+    value_error = False
+    try:
+        decimal = Decimal(n)
+    except InvalidOperation:
```
addressed in [761f54b](https://github.com/dimagi/commcare-hq/pull/31402/commits/761f54b9603672117579b95ba2c51515f87e9cfe)

esoergel committed:
Make data dictionary property types into variables

I've seen this exact type of problem before, but still haven't found a
perfect solution.  I almost want to use an Enum, but accessing the slug
and display are awkward.

esoergel committed:
We only need gps props for now, make that clear

esoergel committed:
stickler

esoergel committed:
Remove old one-off migration and turn into utility

You'll notice the extra level of indirection was removed - I believe
this never actually ran properly.  If it had, it would error - `app` is
not a valid parameter (should be `hqapps`).

esoergel committed:
Automatically apply new case search mapping

snopoke commented on corehq/apps/data_dictionary/models.py:
```diff
@@ -8,18 +9,23 @@
+    UNDEFINED = property_type('', _('No Type Currently Selected'))
+
+    @classmethod
+    def get_all(cls):
```
if you make this class an `enum` you don't need this method since enums are iterable

esoergel replied:
I did that at first, but it makes accessing it a little awkward.  The two options I considered there were:
```python
class PROPERTY_TYPES(Enum):
    gps = _('GPS')  # the field name doubles as the slug, so it needs to be lowercase

> PROPERTY_TYPES.gps.name
'gps'
> PROPERTY_TYPES.gps.value
'GPS'
```
I don't love this option, as `name` and `value` aren't really descriptive, and the "name" for undefined is `""`, so I'd have to work around that, I expect.

```python
class PROPERTY_TYPES(Enum):
    GPS = property_type('gps', _('GPS'))

> PROPERTY_TYPES.GPS.value.slug
'gps'
> PROPERTY_TYPES.GPS.value.display
'GPS'
```

Farid pointed me to https://docs.djangoproject.com/en/4.0/ref/models/fields/#field-choices-enum-types, which seems appropriate here, though access is the same as in my first example above.  It's not available in our version of Django, but I could be convinced to switch to that first option if that's a pattern we should expect to become familiar with.

snopoke reviewed (approved):
no blocking feedback

esoergel committed:
typo

joxl commented on corehq/pillows/mappings/case_search_mapping.json:
```diff
@@ -47,6 +47,9 @@
                     }
+                },
+                "geo_point": {
+                    "type": "geo_point"
```
Is there a forum post or even Elastic doc that describes how `ignore_malformed` is broken on 2.4?

esoergel replied:
Yep, I just jotted down some links I'd looked at here:
https://dimagi-dev.atlassian.net/browse/USH-1753?focusedCommentId=200782
I don't want to link the issues directly from github, as I don't want back-links going from their repo to ours, highlighting the fact that we're on an embarrassingly out-of-date version.

joxl reviewed (approved):
No blockers from me. Admittedly I glossed over a lot of the business logic as my review is mostly around the bits that pertain to the Elasticsearch backend.  Only question I have is about documentation of the `ignore_malformed` bug.
